### PR TITLE
[RPA3] Improve mxu scheduling util to ~100% and support use_causal_mask, skip_kv_mask, out_dtype

### PR DESCRIPTION
**Pull request by bythew3i**
# Description

This PR
- uses strided load trick to remove all relayout
- uses q_type by default for accumulation dtype to reduce VPU pressure
- allows user specify out_dtype 
- allows user optionally use causal mask
- allows user skip kv masking if not use  causal mask and kv_lens[i] % bkv_csz == 0

# Tests

Tested
```
pytest -x tests/kernels/ragged_paged_attention_kernel_v3_test.py 
```

# Benchmarking

* Llama 3-8: 14.53 -> 15.07 (*3.7% Improvement*)
Combine the change from #1820 , Improvement: 12.89  -> 15.07 reqs/s (**17%**)

```
VLLM_USE_V1=1 vllm serve meta-llama/Llama-3.1-8B-Instruct  --disable-log-requests --port 8003 --gpu-memory-utilization 0.95 --max-num-batched-tokens 512 --tensor-parallel-size 1 --max-model-len 2048 --max_num_seqs 512 --download_dir=/mnt/pd/models &> /mnt/pd/tmp/serve.log &

vllm bench serve --backend vllm --model meta-llama/Llama-3.1-8B-Instruct --dataset-name random --port=8003 --random-input-len 1800 --random-output-len 128
```

Before:
```
============ Serving Benchmark Result ============
Successful requests:                     1000      
Failed requests:                         0         
Benchmark duration (s):                  68.81     
Total input tokens:                      1799000   
Total generated tokens:                  128000    
Request throughput (req/s):              14.53     
Output token throughput (tok/s):         1860.15   
Peak output token throughput (tok/s):    2192.00   
Peak concurrent requests:                1000.00   
Total token throughput (tok/s):          28004.00  
---------------Time to First Token----------------
Mean TTFT (ms):                          34021.48  
Median TTFT (ms):                        34011.85  
P99 TTFT (ms):                           67010.21  
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          17.81     
Median TPOT (ms):                        18.00     
P99 TPOT (ms):                           18.08     
---------------Inter-token Latency----------------
Mean ITL (ms):                           17.82     
Median ITL (ms):                         18.09     
P99 ITL (ms):                            20.04     
==================================================
```

After:
```
============ Serving Benchmark Result ============
Successful requests:                     1000      
Failed requests:                         0         
Benchmark duration (s):                  66.35     
Total input tokens:                      1799000   
Total generated tokens:                  128000    
Request throughput (req/s):              15.07     
Output token throughput (tok/s):         1929.02   
Peak output token throughput (tok/s):    2474.00   
Peak concurrent requests:                1000.00   
Total token throughput (tok/s):          29040.79  
---------------Time to First Token----------------
Mean TTFT (ms):                          32791.43  
Median TTFT (ms):                        32772.91  
P99 TTFT (ms):                           64574.31  
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          17.16     
Median TPOT (ms):                        17.33     
P99 TPOT (ms):                           17.42     
---------------Inter-token Latency----------------
Mean ITL (ms):                           17.17     
Median ITL (ms):                         17.38     
P99 ITL (ms):                            19.03     
==================================================
```

# MXU utilization

Before: 
Screenshot 2026-03-19 at 1 39 06 AM

After:
Screenshot 2026-03-19 at 1 39 42 AM

### CODE REVIEW

File: tpu_inference/kernels/ragged_paged_attention/v3/kernel.py [hunk 1]
```diff
@@ -1568,12 +1634,27 @@ def ragged_paged_attention(
...
   """
     q, k, v = queries, keys, values
     tpu_version = get_tpu_version()
+
+    if out_dtype is None:
+        out_dtype = jnp.float32 if q.dtype == jnp.float32 else jnp.bfloat16
+
+    if mask_value is None:
+        # We do not set to -inf directly because (-inf) - (-inf) is nan.
+        mask_value = -float(jnp.finfo(out_dtype).max)
```
Comment (kyuyeunk) [inline tpu_inference/kernels/ragged_paged_attention/v3/kernel.py]: why not just `jnp.finfo(out_dtype).min`?

File: tpu_inference/kernels/ragged_paged_attention/v3/kernel.py [hunk 2]
```diff
@@ -1568,12 +1634,27 @@ def ragged_paged_attention(
...
+
+    if vmem_limit_bytes is None:
+        # TODO(jevinjiang, jacobplatin): change this to use
+        # `get_vmem_estimate_bytes` when VREG spilling is fixed.
+        vmem_limit_bytes = DEFAULT_VMEM_LIMIT_BYTES
```
Comment (kyuyeunk) [inline tpu_inference/kernels/ragged_paged_attention/v3/kernel.py]: nit: use `pltpu.get_tpu_info().vmem_capacity_bytes` instead.

File: tpu_inference/kernels/ragged_paged_attention/v3/kernel.py [hunk 3]
```diff
@@ -1738,7 +1824,9 @@ def run_rpa_kernel(
                 # Paged attention invokes multiple small DMAs for each pages
                 # instead of a single large DMA. Therefore, the overhead of bounds
                 # checking becomes too significant so we disable it.
-                disable_bounds_checks=True,
+                disable_bounds_checks=disable_bounds_checks,
+                # Only set to true if you gurantee there is no race condition.
+                disable_semaphore_checks=disable_semaphore_checks,
```
Comment (kyuyeunk) [inline tpu_inference/kernels/ragged_paged_attention/v3/kernel.py:1821]: this looks a bit scary?

File: tpu_inference/kernels/ragged_paged_attention/v3/kernel.py [hunk 4]
```diff
@@ -1487,11 +1536,15 @@ def get_default_block_sizes(
     }
 
 
-@jax.jit(
+@functools.partial(
```
Comment (kyuyeunk) [inline tpu_inference/kernels/ragged_paged_attention/v3/kernel.py]: nit: use `@jax.jit` instead.

File: tpu_inference/kernels/ragged_paged_attention/v3/kernel.py [hunk 5]
```diff
@@ -1415,10 +1464,10 @@ def _validate_block_sizes(block_sizes, prefix):
     # No constraints for the following inputs.
     del sm_scale
```
Comment (kyuyeunk) [inline tpu_inference/kernels/ragged_paged_attention/v3/kernel.py:1463]: nit: instead use `del sm_scale, mask_value, ...`

File: tpu_inference/kernels/ragged_paged_attention/v3/kernel.py [hunk 6]
```diff
@@ -780,6 +798,31 @@ def strided_load(ref, start, sz, step, *, dtype=None):
             vec = pltpu.bitcast(vec, dtype)
         return vec
 
+    def strided_store(ref, start, sz, step, val):
+        assert get_dtype_packing(ref.dtype) == 1
+        assert ref.dtype == val.dtype
+        assert ref.shape == val.shape
+        assert len(ref.shape) == 2
+        r, l = ref.shape  # noqa
```
Comment (kyuyeunk) [inline tpu_inference/kernels/ragged_paged_attention/v3/kernel.py:804]: nit: just use `l = ref.shape[1]` or `_, l = ref.shape`

**Review by kyuyeunk — COMMENTED** [review]
mostly nit comments.

File: tpu_inference/kernels/ragged_paged_attention/v3/kernel.py (hunk 6 above)
Comment (bythew3i) [reply tpu_inference/kernels/ragged_paged_attention/v3/kernel.py:804]: The `r` is used (line 809). This `noqa` is to disable lint for variable name

File: tpu_inference/kernels/ragged_paged_attention/v3/kernel.py (hunk 5 above)
Comment (bythew3i) [reply tpu_inference/kernels/ragged_paged_attention/v3/kernel.py:1463]: I do not see much difference here. I prefer the current one which is easy to track.

File: tpu_inference/kernels/ragged_paged_attention/v3/kernel.py (hunk 4 above)
Comment (bythew3i) [reply tpu_inference/kernels/ragged_paged_attention/v3/kernel.py]: No rules saying @jax.jit is better here

File: tpu_inference/kernels/ragged_paged_attention/v3/kernel.py (hunk 3 above)
Comment (bythew3i) [reply tpu_inference/kernels/ragged_paged_attention/v3/kernel.py:1821]: Yeah, both `disable_bounds_checks`, `disable_semaphore_checks` are - only turn off them in production for perf

File: tpu_inference/kernels/ragged_paged_attention/v3/kernel.py (hunk 2 above)
Comment (bythew3i) [reply tpu_inference/kernels/ragged_paged_attention/v3/kernel.py]: Not much difference in this case... Mosaic will cap it to correct value as long as it is larger than the cap. But I can change it to use `pltpu.get_tpu_info().vmem_capacity_bytes`

Comment (kyuyeunk) [reply tpu_inference/kernels/ragged_paged_attention/v3/kernel.py]: In general, i'm not a fan of magic numbers as it hurts code readability. most people will probably have no idea where this number came from:`DEFAULT_VMEM_LIMIT_BYTES = 120 * 1024 * 1024`. so changing to `pltpu.get_tpu_info().vmem_capacity_byte` does sound good to me.

File: tpu_inference/kernels/ragged_paged_attention/v3/kernel.py (hunk 1 above)
Comment (bythew3i) [reply tpu_inference/kernels/ragged_paged_attention/v3/kernel.py]: Good point. Idk why I wrote in this weird way ..

File: tpu_inference/kernels/ragged_paged_attention/v3/kernel.py (hunk 2 above)
Comment (bythew3i) [reply tpu_inference/kernels/ragged_paged_attention/v3/kernel.py]: I mean, this is just your preference. The functionality is the same. But anyway as I said, I updated it.

**Review by kyuyeunk — APPROVED** [review]

**Comment by kyuyeunk** [comment]
LGTM! Thanks for the pr!

**Comment by bythew3i** [comment]
> LGTM! Thanks for the pr!

Thanks for the review!!!

File: tpu_inference/kernels/ragged_paged_attention/v3/kernel.py (hunk 4 above)
Comment (bythew3i) [reply tpu_inference/kernels/ragged_paged_attention/v3/kernel.py]: I also updated this.

Comment (kyuyeunk) [reply tpu_inference/kernels/ragged_paged_attention/v3/kernel.py]: thanks!
